servers/quarkus-common/src/main/java/org/projectnessie/quarkus/providers/storage/CacheInvalidationSender.java

```diff
@@ -0,0 +1,290 @@
+              NESSIE_VERSION_STORE_PERSIST,
+              CONFIG_CACHE_INVALIDATIONS_VALID_TOKENS);
+        } else {
+          LOGGER.info(
+              "Sending remote cache invalidations to {}, currently resolved to {}",
+              serviceNames,
+              resolved);
+        }
+      } catch (Exception e) {
+        throw new RuntimeException(
+            "Failed to resolve service names " + serviceNames + " for remote cache invalidations",
+            (e instanceof ExecutionException) ? e.getCause() : e);
+      }
+    } else if (token != null) {
+      LOGGER.warn(
+          "No service names are configured to send cache invalidation messages to - will not send any invalidation message. You need to configure the service name(s) via {}.{}",
+          NESSIE_VERSION_STORE_PERSIST,
+          CONFIG_CACHE_INVALIDATIONS_SERVICE_NAMES);
+    }
+  }
+
+  private Future<List<String>> updateServiceNames() {
+    return resolveServiceNames(serviceNames)
+        .map(all -> all.filter(adr -> !LOCAL_ADDRESSES.contains(adr)).toList())
+        .onSuccess(
+            all -> {
+              // refresh addresses regularly
+              scheduleServiceNameResolution();
+
+              LOGGER.debug("Resolved service names {} to {}", serviceNames, all);
+              resolvedAddresses = all;
+            });
+  }
+
+  private void updateServiceNamesHandleFailure() {
+    updateServiceNames()
+        .onFailure(
+            failure -> {
+              // refresh addresses regularly
+              scheduleServiceNameResolution();
+
+              LOGGER.warn(
+                  "Failed to resolve service names {} for remote cache invalidations",
+                  serviceNames);
+            });
+  }
+
+  private void scheduleServiceNameResolution() {
+    vertx.setTimer(serviceNameLookupIntervalMillis, x -> updateServiceNamesHandleFailure());
+  }
+
+  @VisibleForTesting
+  Future<Stream<String>> resolveServiceNames(List<String> serviceNames) {
+    return addressResolver.resolveAll(serviceNames);
+  }
+
+  void enqueue(CacheInvalidation invalidation) {
+    if (serviceNames.isEmpty() || token == null) {
+      // Don't do anything if there are no targets to send invalidations to or whether no token has
+      // been configured.
+      return;
+    }
+
+    lock.lock();
+    try {
+      invalidations.add(invalidation);
+
+      if (!triggered) {
+        LOGGER.trace("Triggered invalidation submission");
+        vertx.executeBlocking(this::sendInvalidations);
+        triggered = true;
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Void sendInvalidations() {
+    while (true) {
+      List<CacheInvalidation> batch = new ArrayList<>(batchSize);
+      lock.lock();
+      try {
+        invalidations.drainTo(batch, 100);
+        if (batch.isEmpty()) {
+          LOGGER.trace("Done sending invalidations");
+          triggered = false;
```

Comment:
If something throws in this loop (e.g. `submit()`), `triggered` will never be reset it seems, and the sending of notifications will never resume... or did I misinterpret? Is that a concern?

Reply:
I think that's not really a concern. But I've added something around it.